src/Two/AbstractProvider.php

```diff
@@ -434,6 +456,49 @@ protected function getState()
         return Str::random(40);
     }
 
+    /**
+     * Determine if the provider uses PKCE.
+     *
+     * @return bool
+     */
+    protected function usesPKCE()
+    {
+        return $this->usesPKCE;
+    }
+
+    /**
+     * Generates a random string of the right length for the PKCE code verifier.
+     *
+     * @return string
+     */
+    protected function getCodeVerifier()
+    {
+        return Str::random(96);
+    }
+
+    /**
+     * Generates the PKCE code challenge based on the PKCE code verifier in the session.
+     *
+     * @return string
+     */
+    protected function getCodeChallenge()
+    {
+        $verifier = $this->request->session()->pull('code_verifier');
```

Comment:
If you pull the value from the session here, the `code_verifier` is no longer available for validation (on line 312)

Reply:
Thanks, I changed it to a `get`